### PR TITLE
Export Julia's LLVM passes using LLVM C API compatible functions.

### PR DESCRIPTION
**Pull request by maleadt**
Used for [CUDAnative.jl](https://github.com/JuliaGPU/CUDAnative.jl/blob/14f215aac8f524094d69b5185b2976a856dcbe86/src/jit.jl#L214-L230).

### CODE REVIEW

**Comment by vtjnash** [comment]
Maybe would be easier to just eliminate these tbaa node arguments?

**Comment by yuyichao** [comment]
> Maybe would be easier to just eliminate these tbaa node arguments?

Sounds good to me, I initially did this to make it as independent as possible but since I couldn't figure out a way to make the pass truly independent of the context it probably not worth doing this anymore...

**Comment by maleadt** [comment]
Eliminated the tbaa node arguments, recreating them in the pass. Did so in `runOnModule`, at the cost of some extra verbosity, to have it working with multiple contexts.

Should be identical, but I guess it can't hurt:

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19024#issuecomment-254934925) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/05691e7399cda0d625839e6899c27b151c431361/c0feabd_vs_0faf8ce/report.md). cc @jrevels

**Comment by maleadt** [comment]
OK, looks good. With Travis OSX being dead, I built and tested on the MIT macmini which went fine.

File: src/llvm-gcroot.cpp [hunk 1]
```diff
@@ -1175,6 +1176,10 @@ static void ensure_enter_function(Module &M)
 
 bool LowerGCFrame::runOnModule(Module &M)
 {
+    MDBuilder *mbuilder = new MDBuilder(M.getContext());
```
Comment (yuyichao) [inline src/llvm-gcroot.cpp]: Should this be stack allocated instead?

Comment (maleadt) [inline src/llvm-gcroot.cpp]: Probably yeah, I copied without paying attention...

**Review by yuyichao — APPROVED** [review]
LGTM assuming you've check that the child node we create each time are the same one. (so that we are not creating a increasing numbers of them).

**Comment by maleadt** [comment]
@yuyichao Yeah I verified, no new nodes are created. But even if they were, they should 'alias' I guess (`may be identified (uniqued) with other roots with the same name`).

I'll merge if the board turns green.

File: src/codegen.cpp [hunk 2]
```diff
@@ -5078,7 +5085,7 @@ static std::unique_ptr<Module> emit_function(jl_method_instance_t *lam, jl_code_
 
 // --- initialization ---
 
-static MDNode *tbaa_make_child( const char *name, MDNode *parent, bool isConstant=false )
+MDNode *tbaa_make_child(const char *name, MDNode *parent, bool isConstant=false)
 {
     MDNode *n = mbuilder->createTBAANode(name,parent,isConstant);
```
Comment (vtjnash) [inline src/codegen.cpp]: `MDBuilder mbuilder(parent->getContext())`

the `MDBuilder` class is just a single pointer (to a LLVMContext), it doesn't need to be a global

**Comment by vtjnash** [comment]
LGTM, removing globals from codegen is a future cleanup action and doesn't need to be done now

**Pull request merged**